### ISSUE TITLE
PackageInstallerService.ProcessProjectChange is sending non-fatal watsons causing UI delays

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#20101]
Investigating UI delays in https://developercommunity.visualstudio.com/content/problem/66362/visual-studio-2017-keeps-freezing-on-me.html, I see a bunch are proceeded immediately by the PackageInstallerService sending on non-fatal watsons on UI thread for the following:

```
System.ArgumentException: An item with the same key has already been added.
 at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
 at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
 at System.Collections.Generic.Dictionary`2.System.Collections.Generic.ICollection<System.Collections.Generic.KeyValuePair<TKey,TValue>>.Add(KeyValuePair`2 keyValuePair)
 at Roslyn.Utilities.ICollectionExtensions.AddRange[T](ICollection`1 collection, IEnumerable`1 values)
 at Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.ProcessProjectChange(Solution solution, ProjectId projectId)
 --- End of stack trace from previous location where exception was thrown ---
 at 
```

I'm guessing this line is failing: 
https://github.com/dotnet/roslyn/blob/cdde5bd4d36d06b7f8ebe434e82eb5e35b4e1277/src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs#L462.

**Comment by davkean** [comment]
Looks like we sent 41 non-fatal watsons in that session - shouldn't these be throttled?

**Comment by davkean** [comment]
tag @CyrusNajmabadi

**Comment by onovotny** [comment]
If it helps, I was working on this solution: https://github.com/AArnott/Nerdbank.GitVersioning. I do not have the `NuProj` Addin installed, so I did get an unsupported warning about the NuProj project not being supported. 

I got the freezes while typing both in C# code and editing json files (version.schema.json).

**Comment by Pilchie** [comment]
I think that was fixed recently.  Also, what's causing a UI delay about that? Just walking the callstack?  AFAIK, we don't actually much more than that in the NFW codepath.

cc @tmeschter and @heejaechang though.

**Comment by heejaechang** [comment]
according to telemetry team, they do throttle. let me add in team that owns NFW.

**Comment by CyrusNajmabadi** [comment]
Roslyn exception fixed with https://github.com/dotnet/roslyn/pull/19894

**Comment by Pilchie** [comment]
Assigning to @heejaechang to follow up on throttling in general, since this particular exception is fixed.

**Comment by heejaechang** [comment]
I sent email to Stephen who explained about throttle to us before. I pinged him again

**Comment by davkean** [comment]
@Pilchie If you look at the events, there's a clear correlation between sending a watson and the UI delay end marker. I've personally run into this. where capturing a non-fatal watson visible blocks the UI.

**Comment by heejaechang** [comment]
NFW getting dump which freeze the process I can understand, what I dont understand is it happening multiple times. since Stephen told us most likely user won't see it happening multiple times for same issue.

**Comment by lordofstuff** [comment]
Hey, Stephen here. I work on the Fault Event API frequently. We do throttle to 10% of fault events sending to Watson, and 10 of those per session. We do not throttle per issue, so if a user hits the same issue 41 times in a session, it's entirely reasonable that they will send four reports during that session. If you want to throttle it per issue, I can help you with that, and we might consider implementing something similar natively in the API. 

But as of right now, this sounds like it is working by design, and the bug should be looking into why this event fires so frequently.

**Comment by Pilchie** [comment]
From talking to @tmeschter though, we aren't actually collecting those dumps, just the callstack in the telemetry event...

**Comment by Pilchie** [comment]
Or are we collecting them, and then just failing to upload them? 😢

**Comment by heejaechang** [comment]
@Pilchie that is fixed with https://github.com/dotnet/roslyn/pull/19787

**Comment by Pilchie** [comment]
But @davkean is seeing this on existing public bits from 15.2.

**Comment by heejaechang** [comment]
@lordofstuff @Pilchie then, it is a question to NFW team. Stephen, I think you guys told us that unless we explicitly add us to the list, the API won't collect dump. is that not true?

**Comment by lordofstuff** [comment]
What machines/users it collects dumps from is entirely out of the control of both the Faut event API and consumers of the API. WER itself decides whether it should collect several dumps from the same machine or from a wide variety of sources, or if it has collected enough and does not need any more. 

It is true that it should not be collecting a dump if you did not specifically request one, but even attaching other files or sending a report at all can still be expensive, especially if these 41 events were tightly clustered.

**Comment by davkean** [comment]
Some of the events were clustered together, but for the most part they were not. Whatever it was doing, it was trigging the UI delay timeout - what is that set to? 2 seconds?

**Comment by lordofstuff** [comment]
The UI Delay threshold is 1 second, and it's perfectly normal for a Watson report to take that long.

**Comment by heejaechang** [comment]
hmmm.. hard to believe NFW would take over 1 second without gathering dump. let me take a look at the kusto.

**Comment by lordofstuff** [comment]
A few things to keep in mind: 
This occurs in the middle of other CPU work, so there may be other things contributing to the delay before and after the NFW is submitted.
WER does file I/O, launches a new process, and makes network calls while submitting a report.
It freezes all threads and does not pump any messages while the report is being submitted. 

This all means that the time can also vary based on disk speed, network speed, and other things out of our control.

**Comment by heejaechang** [comment]
@lordofstuff how do I know NFW created a dump or not from telemetry event?

**Comment by heejaechang** [comment]
@davkean @lordofstuff I dont know whether dump is created or not, but I see 6 sampled event among +-70 NFW events. so regardless of dump, sending watson looks like expansive.

**Comment by MattGertz** [comment]
Mondo escrow triage:  moving this out; let's keep the investigation going though, and tag me if there is something actionable that would be worth bringing in Escrow.

**Comment by heejaechang** [comment]
I am closing this. if it repro, let me know.

**Comment by davkean** [comment]
We're still hits of this in session data.

**Comment by heejaechang** [comment]
@davkean do you have a dump? or Watson that I can take a look? I am not sure where that AddRange came from.

**Comment by davkean** [comment]
@heejaechang There is no dump, only session data.

**Comment by davkean** [comment]
My guess is two packages with the same name but different version.

**Comment by heejaechang** [comment]
but code you are referring - https://github.com/dotnet/roslyn/blob/cdde5bd4d36d06b7f8ebe434e82eb5e35b4e1277/src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs#L462

uses MultiDictionary which means it allows multiple value for same key. 

so either this is already fixed or actual problem is happening some place else. probably need dump to know sure.

**Comment by davkean** [comment]
You should look at the sessions to see if it's hit past 15.4, Cyrus fixed for that release: https://github.com/dotnet/roslyn/commit/082f9937812a93c339210a7a02e61e73dd4beffc.

**Comment by heejaechang** [comment]
"We're still hits of this in session data." @davkean what session data?

**Comment by davkean** [comment]
Called out in the threads we have with the feedback team.

**Comment by heejaechang** [comment]
@davkean that is different issue than this. that issue is nuget throwing not us. more specifically. this.
var installedPackageMetadata = _packageServices.GetInstalledPackages(dteProject);

service itself is our wrapper on IVsPackageInstallerServices, but regardless, it is nuget throwing. not us for the feedback issue.

**Comment by davkean** [comment]
@heejaechang To avoid this being dropped, please confirm that we have stopped throwing these exceptions or a bug has been filed against NuGet. Do not close this bug until we've confirmed it has been fixed, it's one of the top NFW.

**Comment by heejaechang** [comment]
@davkean hmmm?? that is different issue? I am not sure why you connect that issue with this one?

**Comment by davkean** [comment]
Ah my bad, I was under the impression it was the same. Did we file a bug against NuGet for the later issue?

**Comment by heejaechang** [comment]
@davkean I didn't check. tagging people who deal with NFW and triaging. @jinujoseph @davkean @ivanbasov

**Comment by davkean** [comment]
I'm still running into this issue. I filed https://github.com/dotnet/project-system/issues/3825 to track reporting the watson on UI thread causing a UI delay. I'm reopening this bug because I'm still clearly hitting PackageInstallerService NFW in recent builds. I don't have a repro, just looking through my delays:

```
-- | -- | --
ROOT | 1 | 49,444
ntdll.dll!_RtlUserThreadStart | 1 | 49,444
ntdll.dll!__RtlUserThreadStart | 1 | 49,444
kernel32.dll!BaseThreadInitThunk | 1 | 49,444
devenv.exe!__scrt_common_main_seh | 1 | 49,444
devenv.exe!WinMain | 1 | 49,444
devenv.exe!CDevEnvAppId::Run | 1 | 49,444
devenv.exe!util_CallVsMain | 1 | 49,444
msenv.dll!VStudioMain | 1 | 49,444
msenv.dll!VStudioMainLogged | 1 | 49,444
msenv.dll!CMsoComponent::PushMsgLoop | 1 | 49,444
msenv.dll!SCM_MsoCompMgr::FPushMessageLoop | 1 | 49,444
msenv.dll!SCM::FPushMessageLoop | 1 | 49,444
msenv.dll!CMsoCMHandler::FPushMessageLoop | 1 | 49,444
msenv.dll!CMsoCMHandler::EnvironmentMsgLoop | 1 | 49,444
msenv.dll!MainMessageLoop::ProcessMessage | 1 | 49,444
user32.dll!DispatchMessageW | 1 | 49,444
user32.dll!DispatchMessageWorker | 1 | 49,444
user32.dll!UserCallWinProcCheckWow | 1 | 49,444
user32.dll!_InternalCallWinProc | 1 | 49,444
windowsbase.dll!MS.Win32.HwndSubclass.SubclassWndProc | 1 | 49,444
windowsbase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl | 1 | 49,444
windowsbase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen | 1 | 49,444
windowsbase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall | 1 | 49,444
windowsbase.dll!MS.Win32.HwndSubclass.DispatcherCallbackOperation | 1 | 49,444
windowsbase.dll!MS.Win32.HwndWrapper.WndProc | 1 | 49,444
windowsbase.dll!System.Windows.Threading.Dispatcher.WndProcHook | 1 | 49,444
windowsbase.dll!System.Windows.Threading.Dispatcher.ProcessQueue | 1 | 49,444
windowsbase.dll!System.Windows.Threading.DispatcherOperation.Invoke | 1 | 49,444
windowsbase.dll!MS.Internal.CulturePreservingExecutionContext.Run | 1 | 49,444
mscorlib.dll!System.Threading.ExecutionContext.Run | 1 | 49,444
mscorlib.dll!System.Threading.ExecutionContext.RunInternal | 1 | 49,444
windowsbase.dll!System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext | 1 | 49,444
windowsbase.dll!System.Windows.Threading.DispatcherOperation.InvokeImpl | 1 | 49,444
windowsbase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen | 1 | 49,444
microsoft.codeanalysis.editorfeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.SynchronizationContextTaskScheduler.PostCallback | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.TaskScheduler.TryExecuteTask | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal | 1 | 49,444
mscorlib.dll!System.Threading.ExecutionContext.Run | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.ExecutionContextCallback | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.Execute | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService+<>c__DisplayClass_0.<OnWorkspaceChanged>b__ | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.ProcessBatchedChangesOnForeground | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.ProcessProjectChange | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.CodeAnalysis.ErrorReporting.WatsonReporter.Report | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.TelemetrySession.PostEvent | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.SessionChannel.WatsonSessionChannel.PostEvent | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.PostWatsonReport | 1 | 49,444
mscorlib.dll!System.Threading.WaitHandle.WaitOne | 1 | 49,444
mscorlib.dll!System.Threading.WaitHandle.InternalWaitOne | 1 | 49,444
clr.dll!WaitHandleNative::CorWaitOneNative | 1 | 49,444
clr.dll!Thread::DoAppropriateWait | 1 | 49,444
clr.dll!Thread::DoAppropriateWaitWorker | 1 | 49,444
clr.dll!Thread::DoSyncContextWait | 1 | 49,444
clr.dll!MethodDescCallSite::CallTargetWorker | 1 | 49,444
clr.dll!CallDescrWorkerWithHandler | 1 | 49,444
clr.dll!CallDescrWorkerInternal | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.NoPumpSyncContext.Wait | 1 | 49,444
kernelbase.dll!WaitForMultipleObjects | 1 | 49,444
```

If there is an existing NuGet bug, please point me to it. Otherwise, we risk dropping this again.

**Comment by CyrusNajmabadi** [comment]
reporting on the UI thread is def not good.  Is tehre a way to know what this NFW is though?  can we just fix the issue?

**Comment by CyrusNajmabadi** [comment]
@Pilchie Looking at ths, it looks like we're still getting exceptions somewhere, but there's nothing tracking driving a fix to that issue.  Can someone look into that?  It may be affecting lots of customers, and is very likely hte issue for @davkean

**Comment by Pilchie** [comment]
I think this is still assigned to @heejaechang and open... 

Also @jinujoseph

**Comment by heejaechang** [comment]
hmm.. didn't know I had this bug. @jinujoseph

**Comment by davkean** [comment]
We keep closing it, but it's clearly not fixed nor is it dup'd against an existing bug.

**Comment by CyrusNajmabadi** [comment]
@heejaechang Can you check watson and see what's up here?  Thanks!

**Comment by davkean** [comment]
This is the stack:

```
System.InvalidOperationException: The operation failed as details for project ConsoleApp270 could not be loaded.

   at NuGet.PackageManagement.VisualStudio.NetCorePackageReferenceProject.GetPackageSpecsAsync(DependencyGraphCacheContext context)
   at NuGet.VisualStudio.VsPackageInstallerServices.<GetPackagesPathResolverAsync>d__8.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.VisualStudio.VsPackageInstallerServices.<>c__DisplayClass10_0.<<GetInstalledPackages>b__0>d.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()
   at Microsoft.VisualStudio.Threading.JoinableTask`1.CompleteOnCurrentThread()
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run[T](Func`1 asyncMethod, JoinableTaskCreationOptions creationOptions)
   at NuGet.VisualStudio.VsPackageInstallerServices.GetInstalledPackages(Project project)
   at Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.PackageServicesProxy.GetInstalledPackages(Project project)
   at Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.ProcessProjectChange(Solution solution, ProjectId projectId)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

This is because this service is querying the project before "we've nominated" the project so NuGet doesn't know about the project. Can we avoid looking up the project for packages until the first design-time build has finished until we've got a better way to query NuGet as to whether we've notified them.

Note: This _still_ hangs VS due to https://github.com/dotnet/project-system/issues/3825.

**Comment by davkean** [comment]
Or alternatively, change this path to handle "InvalidOperationException".

**Comment by CyrusNajmabadi** [comment]
Could we have things not throw here?  Can tehre just be a 'Try' path that we could call instead?  It seems very strange that an external component would need to know what internal state this other component needs to be in in order to simple query it for data.

**Comment by CyrusNajmabadi** [comment]
> Can you please report the results of any conversations back here. This is important information to understand that affects development in the Roslyn codebase.

I would def not want to do that.  That's a large potential pool of exceptions that would then mean that future throws might not get caught.

**Comment by ryzngard** [comment]
This instance has been fixed by #32933, and remaining work will be under #29100 to move our NFW off the UI thread.  See the PR for more context on the current resolution.
